### PR TITLE
Bugfix for np.interp

### PR DESCRIPTION
**Pull request by rjenc29**
Initial commit for CI.  Details to follow.

### CODE REVIEW

**Comment by rjenc29** [comment]
So this fixes a bug exemplified by the following:

```python
x = np.array([1, 2, 3, 4])
xp = np.array([1, 2, 3, 4])
fp = np.array([1, 2, 3.5, 4])

np.interp(x, xp, fp)
>>> [1.  2.  3.5 4. ]

njit()(lambda *args: np.interp(*args))(x, xp, fp)
>>> [1.  3.  3.5 4. ]  # note second value is 3. but should be 2.
```

Additionally, there were a couple of inconsistencies versus numpy around handling of non finite fp values, which I think I have fixed (and added tests for).

The faulty logic was around the search space reduction, to try to limit the work done by the binary search of xp values.  I have lifted this out.  

Cheers

**Comment by rjenc29** [comment]
PS: a benchmarking script (borrowed and adapted from an old numpy PR) in case it's of interest:
https://gist.github.com/rjenc29/2a13d1ecd34892ec1df68087dff9c383

File: numba/tests/test_np_functions.py [hunk 1]
```diff
@@ -2148,6 +2184,73 @@ def test_interp_exceptions(self):
 
         self.assertIn(complex_dtype_msg, str(e.exception))
 
+    def test_interp_non_finite_calibration(self):
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: Needs: ` @unittest.skipUnless(np_version >= (1, 10), "interp needs Numpy 1.10+")` else https://travis-ci.org/numba/numba/jobs/497823519

File: numba/tests/test_np_functions.py [hunk 2]
```diff
@@ -2148,6 +2184,73 @@ def test_interp_exceptions(self):
...
+        params = {'x': x, 'xp': xp, 'fp': fp}
+        _check(params)
+
+        xp = np.array([-np.inf, 1, 9, np.inf])
+        fp = np.array([0, 0.1, 0.9, 1])
+        x = np.array([0.2, 9.5])
+        params = {'x': x, 'xp': xp, 'fp': fp}
+        _check(params)
+
+    def test_interp_supplemental_tests(self):
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: Needs: ` @unittest.skipUnless(np_version >= (1, 10), "interp needs Numpy 1.10+")` else https://travis-ci.org/numba/numba/jobs/497823519

File: numba/tests/test_np_functions.py [hunk 3]
```diff
@@ -2148,6 +2184,73 @@ def test_interp_exceptions(self):
...
+        np.testing.assert_almost_equal(cfunc(x0, x, y), y0)
+
+        x = np.linspace(0, 1, 5)
+        y = np.linspace(0, 1, 5)
+        x0 = np.array(0.3)
+        np.testing.assert_almost_equal(cfunc(x0, x, y), x0)
+
+        xp = np.arange(0, 10, 0.0001)
+        fp = np.sin(xp)
+        np.testing.assert_almost_equal(cfunc(np.pi, xp, fp), 0.0)
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: Perhaps check e.g. `pi/4` intervals in a defined out of order manner so as to jump around on the curve, this deliberately stressing the issue existing in the code without this patch?

**Comment by scootty1** [comment]
As far as I can see, the slope is always calculated with a backward difference. Is this behavior taken from numpy? Would it be useful to rename the slope for example to "slope_at_index_bw" to make clear that it is calculated with a backward difference?

File: numba/tests/test_np_functions.py (hunk 1 above)
Comment (rjenc29) [reply numba/tests/test_np_functions.py]: Done

File: numba/tests/test_np_functions.py (hunk 2 above)
Comment (rjenc29) [reply numba/tests/test_np_functions.py]: Done

File: numba/tests/test_np_functions.py (hunk 3 above)
Comment (rjenc29) [reply numba/tests/test_np_functions.py]: Done, good call

**Comment by rjenc29** [comment]
@scootty1 the numpy implementation has a few branches, but the slope cache is pretty similar:

```c
for (i = 0; i < lenxp - 1; ++i) {
    slopes[i] = (dy[i+1] - dy[i]) / (dx[i+1] - dx[i]);
}
```

versus:
```python
slopes = (fp_arr[1:] - fp_arr[:-1]) / (xp_arr[1:] - xp_arr[:-1])
```

I'm not sure what you mean about the backward difference - if I take an example where `xp` and `fp` make the shape of the letter W, then the slope cache looks like this:

```python
xp_arr = np.array([0, 1, 2, 3, 4])
fp_arr = np.array([2, 0, 1, 0, 2])

slopes = (fp_arr[1:] - fp_arr[:-1]) / (xp_arr[1:] - xp_arr[:-1])
slopes
>>> array([-2.,  1., -1.,  2.])
```
I guess as long as the slope and `dx` are consistent then it should come out in the wash - but apologies if it's not clear & please do PR if there's something clearer; that would be great.

**Comment by rjenc29** [comment]
@stuartarchibald so I have a change which offers a speedup of around x2, averaged across benchmarks (it's faster in all cases - just some more than others).  

All tests pass locally.

I am happy to commit it, but wouldn't want to keep moving the review goal posts as I appreciate that can be disruptive.

If it's alright with you, I will commit.  Else, I will save it for a later PR.  Let me know.  TA!

**Comment by scootty1** [comment]
@rjenc29 It looks like numpy is also using backwards difference. From my point of view, your implementation looks great. I was just asking out of curiosity.
Backward difference is, when the difference quotient, the slope, `delta_y / delta_x` at point `x_0` is evaluated with `delta_y / delta_x = (f(x_0) - f(x_0 - delta_x)) / delta_x`, meaning the slope **before** the evaluation point is taken. Whereas forward difference is using the slope **after** the evaluated point:  `delta_y / delta_x = (f(x_0 + delta_x) - f(x_0)) / delta_x`.
[This wikipedia article](https://en.wikipedia.org/wiki/Difference_quotient) is explaining some of the basic differencing schemes.
In fact, I was expecting numpy to use something like central difference in the range `xp_arr[1:-1]` and backward/forward differences at the borders.

**Comment by stuartarchibald** [comment]
@rjenc29 thanks for the fixes. The patch series as of 7454888 looks good. I'm happy to re-review with respect to a performance improvement, the implementation in that gist reads relatively easily, we'll just need to watch test coverage of the branches. For reference, 0.43 RC window closes on the 5th March :-) Thanks again.

**Comment by rjenc29** [comment]
@scootty1 OK great, and thanks for the link.

@stuartarchibald very well, all things being equal, I will commit a tweak at the weekend which should offer improved performance in at least some cases and request a re-review.  I appreciate that will be cutting it a bit fine, so happy to go with [7454888](https://github.com/numba/numba/commit/74548883bbce17f32ede447cdcd6f9362d4de6c9) if that ends up being more practical.

**Comment by rjenc29** [comment]
So the last commit seems to offer a reasonable balance between performance and replicating numpy behaviour, I think.

I included the datasets below as unit tests; running them in a test harness to capture performance versus numpy (of course, mileage may vary):

| dataset | numpy (ms) | numba (ms) |
|-|-|-|
| much_finer_grid  | 1.7101 | 0.4843 |
| finer_grid  | 1.7143 | 0.4842 |
| similar_grid  | 1.6924 | 0.484 |
| coarser_grid  | 1.6773 | 0.487 |
| much_coarser_grid  | 1.7492 | 0.4842 |
| finer_stretched_grid  | 1.4062 | 0.4843 |
| similar_stretched_grid  | 1.4008 | 0.4839 |
| finer_compressed_grid  | 1.7909 | 0.4841 |
| similar_compressed_grid  | 1.4843 | 0.4562 |
| warped_grid  | 1.4329 | 0.4564 |
| very_low_noise_grid  | 1.4771 | 0.4566 |
| low_noise_grid  | 1.5026 | 0.4567 |
| med_noise_grid  | 1.4921 | 0.4839 |
| high_noise_grid  | 1.7368 | 0.4844 |
| very_high_noise_grid  | 1.7687 | 0.4841 |
| extreme_noise_grid  | 1.7358 | 0.4563 |
| random_fine_grid  | 1.7167 | 0.4566 |
| random_grid  | 1.7516 | 0.4565 |

There are ways of speeding up certain use cases, but my tests indicate they slow down others (which is probably not surprising). 

The up-front check for `xp` monotonicity (which numpy doesn't do) is also quite a fair wedge of the numba run time.  But I think it makes sense to keep it.

Anyway, let me know what you think.  I will keep an eye on CI over the weekend.

Cheers

**Comment by rjenc29** [comment]
I think the test failures are unrelated (env build errors).

@stuartarchibald please could you re-review if you get a mo - ta!

**Comment by stuartarchibald** [comment]
CI Failures are due to an unrelated conda package constraint solving issue.

**Pull request merged**